xen/arch/x86/tpm.c

```diff
@@ -978,4 +978,136 @@ void __stdcall tpm_extend_mbi(uint32_t *mbi)
     tpm_hash_extend(DRTM_LOC, DRTM_DATA_PCR, (uint8_t *)mbi, *mbi,
                     TXT_EVTYPE_MBI, NULL, 0);
 }
+#else
+static struct slr_entry_policy *tpm_get_policy(void)
+{
+    struct txt_os_mle_data *os_mle;
+    struct slr_table *slrt;
+    struct slr_entry_policy *policy;
+
+    os_mle = txt_os_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+
+    slrt = __va(os_mle->slrt);
+    if ( slrt->magic != SLR_TABLE_MAGIC )
+        panic("SLRT has invalid magic value: %#08x!\n", slrt->magic);
+    /* XXX: are newer revisions allowed? */
+    if ( slrt->revision != SLR_TABLE_REVISION )
+        panic("SLRT is of unsupported revision: %#04x!\n", slrt->revision);
+    if ( slrt->architecture != SLR_INTEL_TXT )
+        panic("SLRT is for unexpected architecture: %#04x!\n",
+                  slrt->architecture);
+    if ( slrt->size > slrt->max_size )
+        panic("SLRT is larger than its max size: %#08x > %#08x!\n",
+                  slrt->size, slrt->max_size);
+    if ( slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_INTEL_INFO) == NULL )
+        panic("SLRT is missing Intel-specific information!\n");
+
+    policy = (struct slr_entry_policy *)
+        slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_DRTM_POLICY);
+    if (policy == NULL)
+        panic("SLRT is missing DRTM policy!\n");
+
+    /* SLRT policy entry must be the first one (so measuring order matches
+     * policy order) and point to SLRT. */
+    /* XXX: are newer revisions allowed? */
+    if ( policy->revision != SLR_POLICY_REVISION )
+        panic("DRTM policy in SLRT is of unsupported revision: %#04x!\n",
+                  slrt->revision);
+    if ( policy->nr_entries == 0 )
+        panic("DRTM policy in SLRT is empty!\n");
+
+    return policy;
+}
+
+static void verify_drtm_mbi_consistency(struct slr_entry_policy *policy,
+                                        struct slr_policy_entry *policy_entry,
+                                        multiboot_info_t *mbi)
+{
+    module_t *mods = __va(mbi->mods_addr);
+    uint32_t i;
+
+    for ( i = 0; i < mbi->mods_count; i++ ) {
```

Comment:
https://github.com/TrenchBoot/xen/compare/bab60f90c52351add03aec69881f03a9c30cda6a..d820e3dac46047e7e81db6fdc30d72041e9d208e#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24aeR1028-R1059

Reply:
Could you reverse the message in `panic()`? Policy in SLRT is the source of truth, and MBI should match it, not the other way around. Current message suggests that policy is wrong in case of mismatch.

Reply:
https://github.com/TrenchBoot/xen/compare/6ec1f7b7ae9af57ca6abbb7729255aaf14fbcfa7..c8f93f26c569ab37177393c2ceedb0cd012da4d1#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24aeL1066-R1063